### PR TITLE
mds,osd,mon: support get|set tcmalloc release rate

### PR DESCRIPTION
**Pull request by Aran85**
Signed-off-by: Zengran Zhang <zhangzengran@sangfor.com.cn>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by Aran85** [comment]
currently, osd will react on admin socket and ceph tell, mds and mon only on ceph tell

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -6208,7 +6209,9 @@ COMMAND("bench " \
 	"osd", "rw")
 COMMAND("flush_pg_stats", "flush pg stats", "osd", "rw")
 COMMAND("heap " \
-	"name=heapcmd,type=CephChoices,strings=dump|start_profiler|stop_profiler|release|stats", \
+	"name=heapcmd,type=CephChoices,strings="\
+	    "dump|start_profiler|stop_profiler|release|(get|set)_release_rate|stats " \
```
Comment (liewegas) [inline src/osd/OSD.cc]: does ceph_argparse understand the (get|set)_release_rate syntax??

**Comment by liewegas** [comment]
aside from the command description, this looks good!

File: src/osd/OSD.cc (hunk 1 above)
Comment (Aran85) [reply src/osd/OSD.cc]: sorry, updated.

File: src/mds/MDSDaemon.cc [hunk 2]
```diff
@@ -838,6 +838,9 @@ int MDSDaemon::_handle_command(
       cmd_getval(cct, cmdmap, "heapcmd", heapcmd);
       vector<string> heapcmd_vec;
       get_str_vec(heapcmd, heapcmd_vec);
+      string value;
+      if(cmd_getval(cct, cmdmap, "value", value))
```
Comment (tchaikov) [inline src/mds/MDSDaemon.cc]: add a space after `if`.

Comment (tchaikov) [inline src/mds/MDSDaemon.cc]: please update `doc/man/8/ceph.rst` accordingly. and might want to note what this option means, for instance,

> TCMalloc releases no-longer-used memory back to the kernel gradually. ``release_rate`` controls how quickly this happens. Increase this setting to make TCMalloc to return unused memory more frequently. 0 means never return memory to system, 1 means wait for 1000 pages after releasing a page to system. It is ``1.0`` by default.

and how is related to "release" command.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -6208,7 +6209,9 @@ COMMAND("bench " \
 	"osd", "rw")
 COMMAND("flush_pg_stats", "flush pg stats", "osd", "rw")
 COMMAND("heap " \
-	"name=heapcmd,type=CephChoices,strings=dump|start_profiler|stop_profiler|release|stats", \
+	"name=heapcmd,type=CephChoices,strings="\
+	    "dump|start_profiler|stop_profiler|release|get_release_rate|set_release_rate|stats " \
+	"name=value,type=CephString,req=false", \
```
Comment (tchaikov) [inline src/osd/OSD.cc]: no need to add a backslash here.

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -6208,7 +6209,9 @@ COMMAND("bench " \
...
 	"show heap usage info (available only if compiled with tcmalloc)", \
```
Comment (tchaikov) [inline src/osd/OSD.cc]: ditto.

File: src/perfglue/heap_profiler.cc [hunk 5]
```diff
@@ -153,6 +163,19 @@ void ceph_heap_profiler_handle_command(const std::vector<std::string>& cmd,
...
+	<< std::setprecision(4) << ceph_heap_get_release_rate() << "\n";
+  } else if (cmd.size() == 2 && cmd[0] == "set_release_rate") {
+    std::string err;
+    double val = strict_strtod(cmd[1].c_str(), &err);
+    if (err.empty()) {
+      ceph_heap_set_release_rate(val);
+      out << g_conf()->name <<  " release rate changed to: " 
+          << std::setprecision(4) << ceph_heap_get_release_rate() << "\n";
+    } else {
+      out << g_conf()->name <<  " *** need an numerical value.";
```
Comment (tchaikov) [inline src/perfglue/heap_profiler.cc]: nit, if you don't care about the **content** of `err`, why bothering using `strict_strtod()`? i mean, `std:: stod()` would suffice here.

**Comment by liewegas** [comment]
retest this please

File: doc/man/8/ceph.rst [hunk 6]
```diff
@@ -323,8 +323,21 @@ Show heap usage info (available only if compiled with tcmalloc)
 
 Usage::
 
-	ceph heap dump|start_profiler|stop_profiler|release|stats
+	ceph heap dump|start_profiler|stop_profiler|stats
 
+Subcommand ``release`` let TCMalloc releases no-longer-used memory back to the kernel gradually.
```
Comment (tchaikov) [inline doc/man/8/ceph.rst]: this is wrong. ``release`` forces TCMalloc to release the memory at once.

File: doc/man/8/ceph.rst [hunk 7]
```diff
@@ -323,8 +323,21 @@ Show heap usage info (available only if compiled with tcmalloc)
...
 
+Subcommand ``release`` let TCMalloc releases no-longer-used memory back to the kernel gradually. 
+
+Usage::
+
+	ceph heap release
+
+Subcommand ``(get|set)_release_rate`` controls how quickly this happens. Increase this setting to
+make TCMalloc to return unused memory more frequently. 0 means never return memory to system, 1 means
+wait for 1000 pages after releasing a page to system. It is `1.0` by default..
```
Comment (tchaikov) [inline doc/man/8/ceph.rst]: please use double quote.

**Comment by Aran85** [comment]
@tchaikov thanks for your patience! updated..

**Pull request merged**